src/hotspot/share/oops/instanceKlass.cpp

```diff
@@ -537,6 +537,17 @@ void InstanceKlass::deallocate_methods(ClassLoaderData* loader_data,
       // Only want to delete methods that are not executing for RedefineClasses.
       // The previous version will point to them so they're not totally dangling
       assert (!method->on_stack(), "shouldn't be called with methods on stack");
+      if (klass) {
+        jmethodID jmid = method->find_jmethod_id_or_null();
+        // Do the pointer maintenance before releasing the metadata, just in case
+        // We need to make sure that jmethodID actually resolves to this method
+        // - multiple redefined versions may share jmethodID slots and if a method
+        //   has already been rewired to a newer version we could be removing reference
+        //   to a still existing method instance
+        if (jmid != nullptr && *((Method**)jmid) == method) {
+          *((Method**)jmid) = nullptr;
+        }
```

Comment:
This should be abstracted behind a utility function e.g. `method->clear_jmethod_id()`.

Reply:
Done